cranelift/frontend/src/frontend.rs

```diff
@@ -610,11 +630,186 @@ impl<'a> FunctionBuilder<'a> {
+    /// live needs-stack-map values at each instruction that must be a safepoint
+    /// (currently this is just non-tail calls).
+    ///
+    /// Second, take those results, add stack slots so we have a place to spill
+    /// to, and then finally spill the live needs-stack-map values at each
+    /// safepoint.
+    fn insert_safepoint_spills(&mut self) {
+        // A map from each safepoint to the set of GC references that are live
+        // across it.
+        let mut safepoints: crate::HashMap<ir::Inst, SmallVec<[ir::Value; 4]>> =
+            crate::HashMap::new();
+
+        // The maximum number of values we need to store in a stack map at the
+        // same time, bucketed by their type's size. This array is indexed by
+        // the log2 of the type's size. We do not support recording values whose
+        // size is greater than 16 in stack maps.
+        let mut max_vals_in_stack_map_by_size = [0, 0, 0, 0, 0];
+
+        // The set of needs-stack-maps values that are currently live in our
+        // traversal.
+        //
+        // NB: use a `BTreeSet` so that iteration is deterministic, as we will
+        // insert spills an order derived from this collection's iteration
+        // order.
+        let mut live = BTreeSet::new();
+
+        // Do our single-pass liveness analysis.
+        //
+        // Use a post-order traversal, traversing the IR backwards from uses to
+        // defs, because liveness is a backwards analysis.
+        //
+        // 1. The definition of a value removes it from our `live` set. Values
+        //    are not live before they are defined.
+        //
+        // 2. When we see any instruction that requires a safepoint (aka
+        //    non-tail calls) we record the current live set of needs-stack-map
+        //    values.
+        //
+        //    We ignore tail calls because this caller and its frame won't exist
+        //    by the time the callee is executing and potentially triggers a GC;
+        //    nothing is live in the function after it exits!
+        //
+        //    Note that this step should actually happen *before* adding uses to
+        //    the `live` set below, in order to avoid holding GC objects alive
+        //    longer than necessary, because arguments to the call that are not
+        //    live afterwards should need not be prevented from reclamation by
+        //    the GC for us, and therefore need not appear in this stack map. It
+        //    is the callee's responsibility to record such arguments in its
+        //    stack maps if it keeps them alive across some call that might
+        //    trigger GC.
+        //
+        // 3. Any use of a needs-stack-map value adds it to our `live` set.
+        //
+        //    Note: we do not flow liveness from block parameters back to branch
+        //    arguments, and instead always consider branch arguments live. That
+        //    additional precision would require a fixed-point loop.
```

Comment:
perhaps to note for more precision here (to guide future thinking around this): would require a fixed-point loop in the presence of backedges, at least?

For example if we (i) do propagate liveness back to branch args from blockparams, and (ii) at a branch, observe that all targets have already been visited (have higher RPO numbers), then we can use the real liveness rather than conservatively assume live.

No real data but my hunch is that this may be important when we have "extraneous blocks" due to Wasm control flow structure that have a single forward out-edge and that we can't yet optimize out (and wouldn't have at this point in the pipeline anyway)...